hamilton/caching/stores/memory.py

```diff
@@ -0,0 +1,279 @@
+from typing import Any, Dict, List, Optional, Sequence
+
+try:
+    from typing import override
+except ImportError:
+    override = lambda x: x  # noqa E731
+
+from hamilton.caching.cache_key import decode_key
+
+from .base import MetadataStore, ResultStore, StoredResult
+from .file import FileResultStore
+from .sqlite import SQLiteMetadataStore
+
+
+class InMemoryMetadataStore(MetadataStore):
```

Comment:
Following our internal discussion, this is not currently possible for `Parallelizable/Collect` scenarios. It would required the [sqlite shared-cache mode](https://www.sqlite.org/sharedcache.html). However, for `python < 3.11`, the standard library `sqlite3` is compiled with the constant `THREADSAFE=1` which makes this feature unusable (or requires the user to recompile SQLite) ([reference](https://ricardoanderegg.com/posts/python-sqlite-thread-safety/))